### DISCUSSION TITLE
Connecting multiple YAMCS together

### DISCUSSION THREAD

**Discussion by codenamefrance** [yamcs/yamcs#987] [Q&A] [answered]
I'm trying to design some sort of distributed telemetry system; the basic idea is to have multiple YAMCS servers (children) and one YAMCS server (father); when a child receives a TM data, I want it to be forwarded to the father server, which will hold all the data of all children servers.

Now, I was wondering if this has already be done, and if there are some pre-built mechanism to achieve this behaviour.
I saw there is a service called 'replication server'; it may work with multiple masters (children) and a slave (father)?

**Comment by xpromache** [comment]
What does the father do with the data? Why can't it receive it directly?

I think you can configure on the father yamcs multiple replication slaves, one for each children:
https://docs.yamcs.org/yamcs-server-manual/services/instance/replication-slave/

You have to pay attention to set lastTxFile different for each.

**Comment by codenamefrance** [comment]
In this idea children are supposed to run on their own in different physical locations. Father's scope is to collect all the data from all children in order to have them in an only location (for tm analysis)

**Reply by xpromache** [reply]
If you want just to collect data, maybe you don't need a full blown Yamcs, in that case I would suggest you have a look at the Yamcs Gateway:
https://github.com/yamcs/yamcs-gateway/tree/main/ygw

But if you want to have a Yamcs for each child, I think replication should work. 

One mistake I made when implementing the replication was to not compress the replication files.  They tend to grow quite large. Compression would be a nice improvement for someone who would want to contribute :)

**Reply by codenamefrance** [reply]
Otherwise, what about writing a script that, when a packet is received, is then sent to the father just like happens between the telemetry source and the child yamcs?

e.g.
X -> telemetry -> YAMCS 1 -> script(telemetry) -> YAMCS 0(father)

In few words, making YAMCS 1 a telemetry source, forgetting for a while that is a yamcs instance?

**Reply by xpromache** [reply, marked as answer]
The replication functionality takes care that if the connection is lost, it is reestablished and it sends the missing packets so the father doesn't lose any data for the children. 

And I think it's not too complicated to setup so you can try it quickly.
